### PR TITLE
Initial version

### PR DESCRIPTION
**Pull request by tomalec**
Starcounter/RebelsLounge#263

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,8 +1,8 @@
 # &lt;enlighted-link&gt;
 
-> A bare minimum custom element starter-kit using [VanillaJS](http://vanilla-js.com/).
->
-> Looking for a working example? Check [hello-world-element](https://github.com/webcomponents/hello-world-element).
+> Custom element to enlight your `<link>`s from Shadow DOM to the Light.
+
+By current spec [HTML Imports does not work in Shadow DOM](https://github.com/w3c/webcomponents/issues/628). That makes it hard to import definitions of custom elements that you use in a shadow root. This element allows you to do that. You could import your dependencies exactly where and when you need them.
```
Comment (alshakero) [inline README.md]: HTML Imports ~~does~~ do not work in Shadow DOM

File: README.md [hunk 2]
```diff
@@ -35,32 +35,27 @@ Or [download as ZIP](https://github.com/Juicy/enlighted-link/archive/master.zip)
 3. Start using it!
 
     ```html
-    <enlighted-link></enlighted-link>
+    <div>
+        #shadow-root
+            <enlighted-link rel="import" href="path/to/some-thing.html"></enlighted-link>
+            <!-- now you can use whatever you imported -->
+            <some-thing></some-thing>
+        #/shadow-root>
```
Comment (alshakero) [inline README.md]: Closing tag isn't right?

File: enlighted-link.html [hunk 3]
```diff
@@ -1,28 +1,39 @@
...
-        constructor() {
-            super();
+        for(let attrNo = 0, len = this.attributes.length; attrNo < len; attrNo++){
+            link.setAttribute(this.attributes[attrNo].name, this.attributes[attrNo].value);
         }
-        // Fires when an instance was inserted into the document
-        connectedCallback() {
-
+        document.head.appendChild(link);
+        // TODO/FIXME(tomalec): attach mutation observer on links parent, to detach this in case link is detached
```
Comment (alshakero) [inline enlighted-link.html]: I don't think you need a Mutation Observer. Just do

`this.link.addEventListener('DOMNodeRemoved', () => this.remove());`

See: https://jsfiddle.net/vqrfewfh/

File: test/added-in-shadow.html [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+        it('once removed should remove related `link` from the document', function() {
+            enlightedLink.remove();
+
+            expect(document.head.querySelector('link[rel="import"][href="assets/doc-with-js.html"]')).to.be.null;
+            expect(document.contains(enlightedLink.link)).to.be.false;
+        });
+    });
+</script>
+
```
Comment (alshakero) [inline test/added-in-shadow.html:52]: If you add the event listener of `DOMNodeRemoved` please add another test here for that case.

File: test/attributes.html [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            enlightedLink.setAttribute('sizes', 'f');
+            enlightedLink.setAttribute('title', 'g');
+
+            expect(link.getAttribute('rel')).to.equal('a');
+            expect(link.getAttribute('href')).to.equal('b');
+            expect(link.getAttribute('media')).to.equal('c');
+            expect(link.getAttribute('hreflang')).to.equal('d');
+            expect(link.getAttribute('type')).to.equal('e');
+            expect(link.getAttribute('sizes')).to.equal('f');
+            expect(link.getAttribute('title')).to.equal('g');
```
Comment (alshakero) [inline test/attributes.html]: I would use `enlightedLink.observedAttributes` array and iterate like

```js
enlightedLink.observedAttributes.forEach(attrib => {
 enlightedLink.setAttribute(attrib, attrib + 'val');
 expect(link.getAttribute(attrib)).to.equal(attrib + 'val');
});
```

Cleaner and easier to maintain.

File: test/attributes.html [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            enlightedLink.removeAttribute('title');
+
+            expect(link.hasAttribute('rel')).to.be.false;
+            expect(link.hasAttribute('href')).to.be.false;
+            expect(link.hasAttribute('media')).to.be.false;
+            expect(link.hasAttribute('hreflang')).to.be.false;
+            expect(link.hasAttribute('type')).to.be.false;
+            expect(link.hasAttribute('sizes')).to.be.false;
+            expect(link.hasAttribute('title')).to.be.false;
+        });
```
Comment (alshakero) [inline test/attributes.html]: Same as above.

**Review by alshakero — CHANGES_REQUESTED** [review]
Nice work. I suggested a few changes.

File: enlighted-link.html (hunk 3 above)
Comment (tomalec) [reply enlighted-link.html]: `MutationEvents` are deprecated in favor of `MutationObserver`. I'd not add feature based on an already deprecated thing. Given MO could be expensive, I'd stick to YAGNI, and just not implement this feature for now. Therefore change `TODO/FIXME` to `IDEA`. WDYT?

**Comment by tomalec** [comment]
I have applied most of your suggestions, except https://github.com/Juicy/enlighted-link/pull/1#discussion_r184648924

There is still one tests failing on Safari https://travis-ci.org/Juicy/enlighted-link/builds/372965215#L1474 I will investigate it once I find a Mac machine to test on.

File: enlighted-link.html (hunk 3 above)
Comment (alshakero) [reply enlighted-link.html]: Good points. And now I read 

"Adding DOM mutation listeners to a document profoundly degrades the performance of further DOM modifications to that document (making them 1.5 - 7 times slower!). Moreover, removing the listeners does not reverse the damage."

from: https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Mutation_events

This is surprising to me!

**Review by alshakero — APPROVED** [review]
LGTM. I fixed the Safari thing.

**Pull request merged**